test/provider/expect.unittest.ts

```diff
@@ -0,0 +1,333 @@
+import 'reflect-metadata';
+
+import {Identity} from 'openfin/_v2/main';
+import {Signal} from 'openfin-service-signal';
+
+import {Model, getId} from '../../src/provider/model/Model';
+import {APIHandler} from '../../src/provider/APIHandler';
+import {APIFromClientTopic} from '../../src/client/internal';
+import {AppDirectory} from '../../src/provider/model/AppDirectory';
+import {Environment} from '../../src/provider/model/Environment';
+import {createMockEnvironmnent, createMockAppWindow} from '../mocks';
+import {Application} from '../../src/client/main';
+import {ContextChannel} from '../../src/provider/model/ContextChannel';
+import {AppWindow} from '../../src/provider/model/AppWindow';
+import {advanceTime, time, useFakeTime, useRealTime} from '../demo/utils/time';
+import {DeferredPromise} from '../../src/provider/common/DeferredPromise';
+import {PartiallyWritable} from '../types';
+import {delay} from '../demo/utils/delay';
+
+jest.mock('../../src/provider/model/AppDirectory');
+jest.mock('../../src/provider/APIHandler');
+
+type TestWindow = {
+    pendingTime?: number,
+    createdTime?: number,
+    connectionTime?: number,
+    closeTime?: number
+    appType: 'directory' | 'non-directory',
+}
+
+type ExpectCall = {
+    callTime: number,
+    finalizeTime: number,
+    result: 'resolve' | 'reject-timeout' | 'reject-closed'
+}
+
+type ExpectCallResult = {
+    promise: Promise<AppWindow>,
+    time: number,
+    call: ExpectCall
+};
+
+type ResultParam = [string, ExpectCall];
+
+type TestParam = [
+    string,
+    TestWindow,
+    number,
+    ExpectCall[]
+];
+
+const FAKE_TEST_DURATION = 10000;
+
+const REGISTRATION_TIMEOUT = 5000;
+const PENDING_TIMEOUT = 100;
+
+let model: Model;
+
+let mockAppDirectory: jest.Mocked<AppDirectory>;
+let mockEnvironment: jest.Mocked<Environment>;
+let mockApiHandler: jest.Mocked<APIHandler<APIFromClientTopic>>;
+
+beforeEach(async () => {
+    mockAppDirectory = new AppDirectory(null!) as jest.Mocked<AppDirectory>;
+    mockEnvironment = createMockEnvironmnent();
+    mockApiHandler = new APIHandler<APIFromClientTopic>() as jest.Mocked<APIHandler<APIFromClientTopic>>;
+    (mockApiHandler as PartiallyWritable<typeof mockApiHandler, 'onConnection'>).onConnection = new Signal<[Identity]>();
+    (mockApiHandler as PartiallyWritable<typeof mockApiHandler, 'onDisconnection'>).onDisconnection = new Signal<[Identity]>();
+
+    model = new Model(mockAppDirectory, mockEnvironment, mockApiHandler);
```

Comment:
Can we use the injector to do this wiring?

eg: https://github.com/HadoukenIO/fdc3-service/blob/develop/test/provider/AppDirectory.unittest.ts#L83-L96

(Will also be simplified a fair bit with SERVICE-593)

Reply:
It doesn't seem to work straightforwardly, and I could spend a lot of time trying to do in a dozen lines what I was doing in one...

Reply:
OK, yeah - maybe a bit verbose without having those changes first. Let's defer to SERVICE-593.